File: Semester 1/Class09_Computer_Contest_Lv_2_Notes_and_Homework_Nov_06-07/lvl2_hw_sol/lvl2_common_words.cpp
#include <iostream>
#include <map>
#include <string>
#include <algorithm>
using namespace std;
typedef map<string, int>::iterator mit;
typedef pair<string, int> psi;
struct WORD{string str; int cnt;}dict2[1002];
int n, m, k; string str, out;
bool cmp(WORD a, WORD b){return a.cnt>b.cnt;}
int main(){
    cin >> n;
    while(n--){
        cin >> m >> k;
        map<string, int> dict;
        for(int i=0; i<m; i++){
            cin >> str;
            dict[str]++;
        }
        if((k % 100)>=11 && (k % 100)<=13) out="th";
        else if(k % 10==1) out="st";
        else if(k % 10 == 2) out="nd";
        else if(k % 10 == 3) out ="rd";
        else out="th";
        cout << k << out <<" most common word(s):\n";
        int tot=0;
        for(mit it=dict.begin(); it!=dict.end(); it++){
            dict2[tot].str = it->first;
            dict2[tot++].cnt = it->second;
        }
        if(k>tot) {cout<<endl; continue;}
        sort(dict2, dict2+tot, cmp);
        if(k==1 || dict2[k-1].cnt<dict2[k-2].cnt){
            for(int i=k-1; dict2[i].cnt==dict2[k-1].cnt && i<tot; i++)
                cout << dict2[i].str << endl;
        }
        cout << endl;
    }
}

File: Semester 1/Class09_Computer_Contest_Lv_2_Notes_and_Homework_Nov_06-07/lvl2_hw_sol/lvl2_common_words.java
import java.util.*;
import java.io.*;
public class homework {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException{
		for(int T = readInt(); T > 0; T--) {
			int n = readInt(), k = readInt();
			Map<String, Integer> map = new HashMap();
			for(int i=0; i<n; i++) {
				String word = next(); map.put(word, map.getOrDefault(word, 0) + 1);
			}
			List<pair> words = new ArrayList();
			for(String key: map.keySet()) words.add(new pair(key, map.get(key)));
			Collections.sort(words);
			String ordinal = getOrdinal(k);
			System.out.println(k + ordinal + " most common word(s):");
			if(k > words.size()) { continue; }
			if(k == 1 || words.get(k-1).cnt < words.get(k-2).cnt) {
				for(int i=k-1; i < words.size() && words.get(i).cnt == words.get(k-1).cnt; i++)
					System.out.println(words.get(i).word);
			}
			System.out.println();
		}
	}
	static String getOrdinal(int k) {
		if(k % 100 >= 11 && k % 100 <= 13) return "th";
		if(k % 10 == 1) return "st";
		if(k % 10 == 2) return "nd";
		if(k % 10 == 3) return "rd";
		return "th";
	}
	static class pair implements Comparable<pair>{
		String word; int cnt;
		pair(String s, int n) { word = s; cnt = n; }
		public int compareTo(pair x) { return -Integer.compare(cnt, x.cnt); }
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}